services/cache/lru.go

```diff
@@ -182,6 +196,34 @@ func (l *LRU) get(key string, value interface{}) error {
 
 		l.evictList.MoveToFront(ent)
 
+		// We use a fast path for hot structs.
+		if msgpVal, ok := value.(msgp.Unmarshaler); ok {
+			_, err := msgpVal.UnmarshalMsg(e.value)
+			return err
+		}
+
+		// This is ugly and makes the cache package aware of the model package.
+		// But this is due to 2 things.
+		// 1. The msgp package works on methods on structs rather than functions.
+		// 2. Our cache interface passes pointers to empty pointers, and not pointers
+		// to values. This is mainly how all our model structs are passed around.
+		// It might be technically possible to use values _just_ for hot structs
+		// like these and then return a pointer while returning from the cache function,
+		// but it will make the codebase inconsistent, and has some edge-cases to take care of.
+		switch v := value.(type) {
+		case **model.User:
+			var u model.User
+			_, err := u.UnmarshalMsg(e.value)
+			*v = &u
+			return err
+		case **model.Session:
+			var s model.Session
+			_, err := s.UnmarshalMsg(e.value)
+			*v = &s
+			return err
+		}
```

Comment:
A couple of question:

1. What was the reason to pass a pointer to a pointer to the `Get` method?
2. (Unrelated to the PR) Do we still need to do deep copies now that we serialize?

Reply:
1. It's OK to just pass a pointer. But the problem is that it only works with normal Marshal/Unmarshal functions. When you have a method of a struct doing the marshal/unmarshalling, then it the problem arises.

This is what we do for example:

```go
var session *model.Session
sessionCache.Get(token, &session)
```

So session is just a `nil` pointer, and then we are populating that pointer variable during unmarshalling. Now since this is essentially a `**`, that means the method set of the type does not satisfy `msp.Unmarshaler`. That's the problem.

2. Nope, I have a PR planned to fix that. Good spot. :)

Reply:
> 1. It's OK to just pass a pointer. But the problem is that it only works with normal Marshal/Unmarshal functions. When you have a method of a struct doing the marshal/unmarshalling, then it the problem arises.
> 
> This is what we do for example:
> 
> ```go
> var session *model.Session
> sessionCache.Get(token, &session)
> ```
> 
> So session is just a `nil` pointer, and then we are populating that pointer variable during unmarshalling. Now since this is essentially a `**`, that means the method set of the type does not satisfy `msp.Unmarshaler`. That's the problem.

I see. I guess the type switch for the fast path could be generated as well at some point if it becomes a burden to maintain.

Reply:
Agree. I don't think we need to add any more structs in the hot path for now. These 2 are fine.

A considerable CPU time is just spent in GC generated from the marshal/unmarshal. So the next steps are to remove the deep copy and try to use a sync.Pool for msgp. And in future modify the other msgpack library to have alloc-free APIs too.